From ae86187f34e2ee53d167e1daacc2025244806356 Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Fri, 29 Jul 2022 09:19:19 +0800
Subject: [PATCH 3657/4585] [Huawei] mm: Use ReliableTotal as upper limit

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187270

--------------------------------

Commit 9e563cc43eba ("[Huawei] mm: Count reliable memory info based on
zone info") count reliable memory based on zone info. However the total
memory between reported by bios and managed by os is different, therefor
use the size reported by bios as upper limit may lead to user confusion
since this value can now seen by user easily.

To fix this confusion, use the size of reliable memory managed by os as its
upper limit.

This value can be seen via cat /proc/meminfo | grep ReliableTotal.

Fixes: 9e563cc43eba ("[Huawei] mm: Count reliable memory info based on zone info")

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 include/linux/mem_reliable.h |  8 ++++----
 mm/mem_reliable.c            | 29 ++++++++++-------------------
 mm/page_alloc.c              |  6 ++++--
 3 files changed, 18 insertions(+), 25 deletions(-)

diff --git a/include/linux/mem_reliable.h b/include/linux/mem_reliable.h
index 28aedeba681e..ddadf2803742 100644
--- a/include/linux/mem_reliable.h
+++ b/include/linux/mem_reliable.h
@@ -25,9 +25,9 @@ extern struct percpu_counter anon_reliable_pages;
 extern unsigned long task_reliable_limit __read_mostly;
 extern atomic_long_t reliable_user_used_nr_page;
 
-extern void add_reliable_mem_size(long sz);
 extern void mem_reliable_init(bool has_unmirrored_mem,
-			      unsigned long *zone_movable_pfn);
+			      unsigned long *zone_movable_pfn,
+			      unsigned long mirrored_sz);
 extern void shmem_reliable_init(void);
 extern void reliable_report_meminfo(struct seq_file *m);
 extern void page_cache_prepare_alloc(gfp_t *gfp);
@@ -141,9 +141,9 @@ static inline void reliable_page_counter(struct page *page,
 
 static inline bool mem_reliable_is_enabled(void) { return false; }
 static inline bool pagecache_reliable_is_enabled(void) { return false; }
-static inline void add_reliable_mem_size(long sz) {}
 static inline void mem_reliable_init(bool has_unmirrored_mem,
-				     unsigned long *zone_movable_pfn) {}
+				     unsigned long *zone_movable_pfn,
+				     unsigned long mirrored_sz) {}
 static inline void shmem_reliable_init(void) {}
 static inline bool skip_none_movable_zone(gfp_t gfp, struct zoneref *z)
 {
diff --git a/mm/mem_reliable.c b/mm/mem_reliable.c
index 7a3ea31e58a3..4540cfc76489 100644
--- a/mm/mem_reliable.c
+++ b/mm/mem_reliable.c
@@ -9,6 +9,8 @@
 #include <linux/mmzone.h>
 #include <linux/crash_dump.h>
 
+#define PAGES_TO_B(n_pages)	((n_pages) << PAGE_SHIFT)
+
 enum mem_reliable_types {
 	MEM_RELIABLE_ALL,
 	MEM_RELIABLE_FALLBACK,
@@ -21,8 +23,6 @@ DEFINE_STATIC_KEY_FALSE(mem_reliable);
 EXPORT_SYMBOL_GPL(mem_reliable);
 
 bool reliable_enabled;
-
-static atomic_long_t total_reliable_mem;
 bool shmem_reliable __read_mostly = true;
 struct percpu_counter reliable_shmem_used_nr_page;
 bool reliable_allow_fallback __read_mostly = true;
@@ -99,16 +99,6 @@ void page_cache_prepare_alloc(gfp_t *gfp)
 	*gfp &= ~GFP_RELIABLE;
 }
 
-void add_reliable_mem_size(long sz)
-{
-	atomic_long_add(sz, &total_reliable_mem);
-}
-
-static unsigned long total_reliable_mem_sz(void)
-{
-	return atomic_long_read(&total_reliable_mem);
-}
-
 static unsigned long total_reliable_pages(void)
 {
 	unsigned long total_reliable_pages = 0;
@@ -138,7 +128,8 @@ static unsigned long used_reliable_pages(void)
 	return total_reliable_pages() - free_reliable_pages();
 }
 
-void mem_reliable_init(bool has_unmirrored_mem, unsigned long *zone_movable_pfn)
+void mem_reliable_init(bool has_unmirrored_mem, unsigned long *zone_movable_pfn,
+		       unsigned long mirrored_sz)
 {
 	if (!reliable_enabled)
 		return;
@@ -148,7 +139,7 @@ void mem_reliable_init(bool has_unmirrored_mem, unsigned long *zone_movable_pfn)
 		return;
 	}
 
-	if (atomic_long_read(&total_reliable_mem) == 0) {
+	if (!mirrored_sz) {
 		memset(zone_movable_pfn, 0,
 		       sizeof(unsigned long) * MAX_NUMNODES);
 		pr_err("init failed, mirrored memory size is zero.\n");
@@ -162,8 +153,7 @@ void mem_reliable_init(bool has_unmirrored_mem, unsigned long *zone_movable_pfn)
 
 	static_branch_enable(&mem_reliable);
 
-	pr_info("init succeed, mirrored memory size(%lu)\n",
-		total_reliable_mem_sz());
+	pr_info("init succeed, mirrored memory size(%lu)\n", mirrored_sz);
 }
 
 void shmem_reliable_init(void)
@@ -222,7 +212,7 @@ int reliable_limit_handler(struct ctl_table *table, int write,
 
 	ret = proc_doulongvec_minmax(table, write, buffer, length, ppos);
 	if (ret == 0 && write) {
-		if (task_reliable_limit > total_reliable_mem_sz() ||
+		if (task_reliable_limit > PAGES_TO_B(total_reliable_pages()) ||
 		    task_reliable_limit <
 			    (task_reliable_used_pages() << PAGE_SHIFT)) {
 			task_reliable_limit = old;
@@ -241,7 +231,8 @@ int reliable_pagecache_max_bytes_write(struct ctl_table *table, int write,
 
 	ret = proc_doulongvec_minmax(table, write, buffer, length, ppos);
 	if (!ret && write) {
-		if (reliable_pagecache_max_bytes > total_reliable_mem_sz()) {
+		if (reliable_pagecache_max_bytes >
+		    PAGES_TO_B(total_reliable_pages())) {
 			reliable_pagecache_max_bytes = old_value;
 			return -EINVAL;
 		}
@@ -336,7 +327,7 @@ int reliable_shmem_bytes_limit_handler(struct ctl_table *table, int write,
 
 	ret = proc_doulongvec_minmax(table, write, buffer, length, ppos);
 	if (!ret && write) {
-		if (*data_ptr > total_reliable_mem_sz()) {
+		if (*data_ptr > PAGES_TO_B(total_reliable_pages())) {
 			*data_ptr = old;
 			return -EINVAL;
 		}
diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index d65dcde1381b..7b591f53dae0 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -7642,10 +7642,11 @@ static void __init find_zone_movable_pfns_for_nodes(void)
 	if (mirrored_kernelcore) {
 		bool mem_below_4gb_not_mirrored = false;
 		bool has_unmirrored_mem = false;
+		unsigned long mirrored_sz = 0;
 
 		for_each_mem_region(r) {
 			if (memblock_is_mirror(r)) {
-				add_reliable_mem_size(r->size);
+				mirrored_sz += r->size;
 				continue;
 			}
 
@@ -7667,7 +7668,8 @@ static void __init find_zone_movable_pfns_for_nodes(void)
 		if (mem_below_4gb_not_mirrored)
 			pr_warn("This configuration results in unmirrored kernel memory.\n");
 
-		mem_reliable_init(has_unmirrored_mem, zone_movable_pfn);
+		mem_reliable_init(has_unmirrored_mem, zone_movable_pfn,
+				  mirrored_sz);
 
 		goto out2;
 	}
-- 
2.17.1

